### PR TITLE
fix: SideCar config in chart

### LINKED ISSUE #3220: Unable to inject sidecar containers in helm charts
### Version

__Karpenter Version:__ v0.20.0

__Kubernetes Version:__ v1.2.4
__Helm Chart Version:__ v0.20.0

I am trying to inject a sidecar container using the helm charts https://github.com/aws/karpenter/tree/main/charts/karpenter there is an option to inject the sidecar containers i.e `controller.sidecarContainer` 

My Relevant values.yaml file is 

```yaml
....
controller:
....
  sidecarContainer:
    image: docker.io/lacework/datacollector:6.2.0-sidecar
    imagePullPolicy: Always
    name: lw-agent-sidecar
    env:
      - name: LaceworkLogStdout
        value: "yes"
      - name: LaceworkAccessToken
        valueFrom:
          secretKeyRef:
            name: lacework-agent-access-token
            key: agent-access-token
      - name: LW_KUBERNETES_NODENAME
        valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
      - name: LW_EXECUTION_ENV
        value: "EKS_FARGATE"
...
```

but I get following error when upgrading helm chart

```bash
Error: UPGRADE FAILED: template: karpenter/templates/deployment.yaml:117:26: executing "karpenter/templates/deployment.yaml" at <.Values.controller.extraVolumeMounts>: nil pointer evaluating interface {}.controller
```

### Expected Behavior

Should be able to add sidecar containers by using the provided helm chart option  i.e `controller.sidecarContainer` 

### Actual Behavior

When the  sidecar  container configs are injected helm chart uprade or install command crashes with following error

```bash
Error: UPGRADE FAILED: template: karpenter/templates/deployment.yaml:117:26: executing "karpenter/templates/deployment.yaml" at <.Values.controller.extraVolumeMounts>: nil pointer evaluating interface {}.controller
```

### Steps to Reproduce the Problem

1. Create new values.yaml file locally
2. add following code snippet
    ```bash
      extraVolumes:
          - name: lacework-volume
            emptyDir: {}
          - name: cfgmap
            configMap:
              name: lacework-config
              items:
              - key: config.json
                path: config.json
          - name: podinfo
            downwardAPI:
              items:
              - path: "labels"
                fieldRef:
                    fieldPath: metadata.labels
              - path: "annotations"
                fieldRef:
                    fieldPath: metadata.annotations
      controller:
          # -- Controller image.
          image: "public.ecr.aws/karpenter/controller:v0.22.1@sha256:8aa720d4fc95d4187e1030f4a11996a27181dada9e6054a77b8ae7b4313d3e6b"
          # -- SecurityContext for the controller container.
          securityContext: {}
          # -- Additional environment variables for the controller pod.
          env: []
          # - name: AWS_REGION
          #   value: eu-west-1
        
          # -- Resources for the controller pod.
          resources:
            requests:
              cpu: 1
              memory: 1Gi
            limits:
              cpu: 1
              memory: 1Gi
          # -- Controller outputPaths - default to stdout only
          outputPaths:
            - stdout
          # -- Controller errorOutputPaths - default to stderr only
          errorOutputPaths:
            - stderr
          # -- Controller log level, defaults to the global log level
          logLevel: ""
          # -- Controller log encoding, defaults to the global log encoding
          logEncoding: ""
          # -- Additional volumeMounts for the controller pod.
          extraVolumeMounts:
          - name: lacework-volume
            mountPath: /var/lib/lacework-backup
          - name: cfgmap
            mountPath: /var/lib/lacework/config
          - name: podinfo
            mountPath: /etc/podinfo
          # - name: aws-iam-token
          #   mountPath: /var/run/secrets/eks.amazonaws.com/serviceaccount
          #   readOnly: true
          # -- Additional sideCarContainer config - this will also inherit volume mounts from deployment
          sidecarContainer:
            image: docker.io/lacework/datacollector:6.2.0-sidecar
            imagePullPolicy: Always
            name: lw-agent-sidecar
            env:
              - name: LaceworkLogStdout
                value: "yes"
              - name: LaceworkAccessToken
                valueFrom:
                  secretKeyRef:
                    name: lacework-agent-access-token
                    key: agent-access-token
              - name: LW_KUBERNETES_NODENAME
                valueFrom:
                    fieldRef:
                      apiVersion: v1
                      fieldPath: spec.nodeName
              - name: LW_EXECUTION_ENV
                value: "EKS_FARGATE"

    ```
3. Run helm upgrade command for karpenter
  ```bash
    helm upgrade --install karpenter -n karpenter oci://public.ecr.aws/karpenter/karpenter -f values.yaml --version v0.22.1 --debug --dry-run
  ```
  
 This is a crash with the following error

```bash
history.go:56: [debug] getting history for release karpenter
upgrade.go:139: [debug] preparing upgrade for karpenter
Error: UPGRADE FAILED: template: karpenter/templates/deployment.yaml:117:26: executing "karpenter/templates/deployment.yaml" at <.Values.controller.extraVolumeMounts>: nil pointer evaluating interface {}.controller
helm.go:88: [debug] template: karpenter/templates/deployment.yaml:117:26: executing "karpenter/templates/deployment.yaml" at <.Values.controller.extraVolumeMounts>: nil pointer evaluating interface {}.controller
UPGRADE FAILED
main.newUpgradeCmd.func2
        helm.sh/helm/v3/cmd/helm/upgrade.go:200
github.com/spf13/cobra.(*Command).execute
        github.com/spf13/cobra@v1.2.1/command.go:856
github.com/spf13/cobra.(*Command).ExecuteC
        github.com/spf13/cobra@v1.2.1/command.go:974
github.com/spf13/cobra.(*Command).Execute
        github.com/spf13/cobra@v1.2.1/command.go:902
main.main
        helm.sh/helm/v3/cmd/helm/helm.go:87
runtime.main
        runtime/proc.go:225
runtime.goexit
        runtime/asm_amd64.s:1371
```

### Resource Specs and Logs

```bash
history.go:56: [debug] getting history for release karpenter
upgrade.go:139: [debug] preparing upgrade for karpenter
Error: UPGRADE FAILED: template: karpenter/templates/deployment.yaml:117:26: executing "karpenter/templates/deployment.yaml" at <.Values.controller.extraVolumeMounts>: nil pointer evaluating interface {}.controller
helm.go:88: [debug] template: karpenter/templates/deployment.yaml:117:26: executing "karpenter/templates/deployment.yaml" at <.Values.controller.extraVolumeMounts>: nil pointer evaluating interface {}.controller
UPGRADE FAILED
main.newUpgradeCmd.func2
        helm.sh/helm/v3/cmd/helm/upgrade.go:200
github.com/spf13/cobra.(*Command).execute
        github.com/spf13/cobra@v1.2.1/command.go:856
github.com/spf13/cobra.(*Command).ExecuteC
        github.com/spf13/cobra@v1.2.1/command.go:974
github.com/spf13/cobra.(*Command).Execute
        github.com/spf13/cobra@v1.2.1/command.go:902
main.main
        helm.sh/helm/v3/cmd/helm/helm.go:87
runtime.main
        runtime/proc.go:225
runtime.goexit
        runtime/asm_amd64.s:1371
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by endriu0**
Fixes #3220

**Description**
Fix sidecar config. Unfortunately did not spend enough time testing it the first time so sending a fixed version now.

This is now tested and included commented out test values. Once uncommented it creates a proper sidecarContainer now and depending if set volumeMounts will contain any volume mounts specified in extraVolumeMounts or sidecarVolumeMounts.

Unfortunately didn't find any nice way of merging the volume mounts without creating a separate value for it.

**How was this change tested?**

* Uncommented all the values for sidecar / helm template -f values.yaml .
* See sidecar config making it to the chart
* See volume mounts when extraVolumeMounts uncommented added to the sidecar VolumeMounts

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note
NONE
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: charts/karpenter/values.yaml [hunk 1]
```diff
@@ -123,8 +123,39 @@ controller:
   # - name: aws-iam-token
   #   mountPath: /var/run/secrets/eks.amazonaws.com/serviceaccount
   #   readOnly: true
-  # -- Additional sideCarContainer config - this will also inherit volume mounts from deployment
-  sidecarContainer: {}
+  # -- Additional sideCarContainer config
+  sidecarContainer: []
+  # - name: fluent-bit
```
Comment (ellistarn) [inline charts/karpenter/values.yaml:128]: Can you remove this from the values?

File: charts/karpenter/values.yaml [hunk 2]
```diff
@@ -123,8 +123,39 @@ controller:
...
+  #     httpGet:
+  #       path: /
+  #       port: http
+  #   readinessProbe:
+  #     httpGet:
+  #       path: /api/v1/health
+  #       port: http
+  # -- Additional volumeMounts for the sidecars - this will be added to the volume mounts on top of extraVolumeMounts
+  sidecarVolumeMounts: []
+  # - mountPath: /fluent-bit/etc/fluent-bit.conf
```
Comment (ellistarn) [inline charts/karpenter/values.yaml:145]: Likewise

File: charts/karpenter/README.md [hunk 3]
```diff
@@ -40,7 +40,8 @@ helm upgrade --install --namespace karpenter --create-namespace \
 | controller.outputPaths | list | `["stdout"]` | Controller outputPaths - default to stdout only |
 | controller.resources | object | `{"limits":{"cpu":1,"memory":"1Gi"},"requests":{"cpu":1,"memory":"1Gi"}}` | Resources for the controller pod. |
 | controller.securityContext | object | `{}` | SecurityContext for the controller container. |
-| controller.sidecarContainer | object | `{}` | Additional sideCarContainer config - this will also inherit volume mounts from deployment |
+| controller.sidecarContainer | list | `[]` | Additional sideCarContainer config |
```
Comment (Flydiverny) [inline charts/karpenter/README.md:43]: 💅 inconsistent casing
```suggestion
| controller.sidecarContainer | list | `[]` | Additional sidecarContainer config |
```

Comment (endriu0) [reply charts/karpenter/README.md:43]: Good catch. 👍

File: charts/karpenter/values.yaml (hunk 1 above)
Comment (endriu0) [reply charts/karpenter/values.yaml]: Done now. Thanks

File: charts/karpenter/values.yaml (hunk 2 above)
Comment (endriu0) [reply charts/karpenter/values.yaml]: 👍

**Review by ellistarn — APPROVED** [review]

**Pull request merged**